### PR TITLE
Search backend: handle `repo:contains` predicates during repo resolution

### PR DESCRIPTION
**Pull request by camdencheek**
This updates our repo resolution to handle `repo:contains.file(x)`, `repo:contains.content(y)`, `repo:contains(file:x content:y)`, and `repohasfile:x`.

Comments inline, but the summary is: 
- `RepoOptions` contains everything we need to know about what repos to search
- The logic in `repo_has_file.go` has been moved into the repo resolution step
- Every combination of file and content filters is represented by the `RepoHasFileContentArgs` struct, which has optional fields
- This works efficiently, and we can now reasonably do `AND`/`OR` queries with `repo:contains...`

Fixes https://github.com/sourcegraph/sourcegraph/issues/23566
Fixes https://github.com/sourcegraph/sourcegraph/issues/20337
Replaces https://github.com/sourcegraph/sourcegraph/pull/37637
Fixes https://github.com/sourcegraph/sourcegraph/issues/26340
Fixes https://github.com/sourcegraph/sourcegraph/issues/19688

~I plan to wait to merge this until after the release cut. It's a somewhat large change, and I'd prefer to give it a release cycle in the wild.~ Release cut has happened. Will merge when approved.

Don't believe the lines added. It's mostly a generated mock and associated tests. This actually removes a couple hundred lines of non-generated code.

## Test plan

[Solid](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@743ce06a9891015caa1b209158616938eca4b4b0/-/blob/dev/gqltest/search_test.go?L1091-1180) backend integration tests, existing unit tests, and added some heavily-mocked tests that hit the partitioning logic a bit. I've done quite a bit of manual edge-case testing as well.

### CODE REVIEW

File: internal/search/repos/repos.go [hunk 1]
```diff
@@ -483,6 +511,237 @@ func (r *Resolver) filterHasCommitAfter(
...
+				rev string
+			}
+			var revsMatchingAllPredicates Set[repoAndRev]
+			for i, opt := range op.HasFileContent {
+				q := zoektQueryForFileContentArgs(opt, op.CaseSensitiveRepoFilters)
+				q = zoektquery.NewAnd(&zoektquery.BranchesRepos{List: indexed.BranchRepos()}, q)
+
+				foundRevs := Set[repoAndRev]{}
+				onMatch := func(res *zoekt.SearchResult) {
+					for _, file := range res.Files {
```
Comment (rvantonder) [inline internal/search/repos/repos.go:608]: wow are you seeing this `range` keyword isn't syntax highlighted? Also the `for` and `if` keywords below? I think your PR broke GH

File: internal/search/types.go [hunk 2]
```diff
@@ -326,6 +310,10 @@ type RepoOptions struct {
 	Limit       int
 	Cursors     []*types.Cursor
 
+	// Whether we should depend on Zoekt for resolving repositories
+	UseIndex       query.YesNoOnly
+	HasFileContent []query.RepoHasFileContentArgs
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/types.go:315]: yes! great.

File: internal/search/repos/repos.go [hunk 3]
```diff
@@ -483,6 +511,237 @@ func (r *Resolver) filterHasCommitAfter(
...
+	}
+
+	foundMatches := false
+	onMatches := func(fms []*protocol.FileMatch) {
+		if len(fms) > 0 {
+			foundMatches = true
+		}
+	}
+
+	_, err := searcher.Search(
```
Comment (rvantonder) [inline internal/search/repos/repos.go:728]: I'm sure there's a good reason, and couldn't easily tell why, but what's stopping us from (re)using backend jobs like [`TextSearchJob`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e92d876e1c6beae6084e65aa536853cdf0231950/-/blob/internal/search/searcher/search.go?L29:6)

**Review by rvantonder — APPROVED** [review]
I'll have to nominate a new PR of the week. Or just give pre-assign it this to next week :P

File: internal/search/repos/repos.go (hunk 3 above)
Comment (camdencheek) [reply internal/search/repos/repos.go:728]: We technically could. However, we need to search each repo/rev individually to know whether they each return results, so all the overhead of the goroutines spun up by the job is unnecessary. Also, I just thought it was easier to read 🤷

Comment (rvantonder) [reply internal/search/repos/repos.go:728]: gotcha, makes sense! just thought all the unused values in this call looked odd but I see this call is actually used deeper in our stack (past `searchFilesInRepos`) makes sense to just go for the direct invocation

**Pull request merged**